[Jetpack App Migration] Update App Banner Branding

SiobhyB opened the pull request:
_Context:_ Following the plans to [refocus the WordPress apps on core features](https://make.wordpress.org/mobile/2022/07/27/refocusing-the-wordpress-app-on-core-features/), we plan to move features that are specific to WordPress.com to the Jetpack app. As such, we will eventually replace all mentions of the WordPress app in Calypso with Jetpack. 

#### Proposed Changes

This PR updates the app banners that are displayed in the Reader, Notifications, Stats, and editor via the mobile web. References to the WordPress app are updated to Jetpack, and the opportunity has also been taken to update the banner's design.

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/187409110-7b5ed896-6674-4228-8b18-12e8bc42083f.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/187408866-8e107c31-8b2c-428d-b8cb-15214e428f9e.png" width="100%"> |

Note, the banner's links won't lead to the Jetpack app with this iteration. This PR is mainly focused on updating the banner's design and the groundwork necessary for the links to work in the future.

I've given an overview of the code changes in the sections below (click each section's heading for more).

<details>
<summary>Banner UI changes ⤵️</summary>

* The `jetpack/app-branding` feature flag (added in https://github.com/Automattic/wp-calypso/pull/66814) is used to add a new banner that displays when the flag is set to true (currently only for wpcalypso and development builds).
* Note, the feature flag, along with any code related to the WordPress-specific banner, will be removed when we're ready to deploy all web-related changes as part of the Jetpack app migration.
* The [new `jetpackAppBanner` component](https://github.com/Automattic/wp-calypso/blob/0eb4baf11960af3d6eaaab2657d8b1246e56e66a/client/blocks/app-banner/index.jsx#L157-L195) is similar to the current `wordpressAppBanner` component, except with changes to the copy, the inclusion of [the `BannerIcon` component](https://github.com/Automattic/wp-calypso/blob/0eb4baf11960af3d6eaaab2657d8b1246e56e66a/client/blocks/app-banner/index.jsx#L245-L263), and the addition of the `jetpack` class name for styling. 
* The `jetpack` class name is currently used to specifically target UI elements in the `jetpackAppBanner`, but will eventually be removed (along with other styles specific to `wordpressAppBanner`) when the changes are ready to be shipped to users.
</details>

<details>
 <summary> Icon animations ⤵️</summary>

* [The `lottie-web` dependency](https://github.com/airbnb/lottie-web) has been added to the `package.json` and `yarn.lock` files. This dependency enables the usage of After Effects animations. As the full features of the library aren't used in our case, we're able to [import the features we need](https://github.com/Automattic/wp-calypso/blob/0eb4baf11960af3d6eaaab2657d8b1246e56e66a/client/blocks/app-banner/index.jsx#L6) from `lottie-web/build/player/lottie_light` for better performance.
* A [new `animations/app-promo` directory](https://github.com/Automattic/wp-calypso/tree/0eb4baf11960af3d6eaaab2657d8b1246e56e66a/static/animations/app-promo) has been added to house the animated icons. These include RTL versions of the icons.
* References to the files have been added to [the `getAppBannerData()` function](https://github.com/Automattic/wp-calypso/pull/67013/files#diff-f48a99fd52f8aa5e35fffd22c9b171f80189818672e89bbc5b2b4ea9d2698257), where they can then be retrieved for use in the banner, depending on the section of the screen the user is on.
* By default, `jest` doesn't handle canvas properly, which leads to an error in tests when combined with `lottie-web`. To resolve this, [`canvas`](https://www.npmjs.com/package/canvas) has also been added as a `devOnly` dependency. 

</details>

<details>
<summary>Deep linking ⤵️</summary>

* **iOS:** As per https://github.com/Automattic/wp-calypso/pull/26118, the app banner points to `apps.wp.com` for iOS users. The Jetpack app won't open until we make updates to that site.
* **Android:** Updates to the URIs have been made [here](https://github.com/Automattic/wp-calypso/blob/9f44d20087bbbb2d3e46417bee9f78ce4babd1fb/client/blocks/app-banner/index.jsx#L125-L143) but won't work until some further changes have been made in the Android codebase (PR pending).
</details>

#### Testing Instructions

_Prerequisite:_ The app banner can be viewed via the mobile web (which can also be [emulated using browser tools](https://developer.chrome.com/docs/devtools/device-mode/#viewport)) when navigating directly to the Reader, Stats, Notifications, or the editor. Note, if the banner has previously been dismissed in your browser, it'll be necessary to clear your cache. 

The following should be confirmed for both iOS and Android:

* Verify that the banner appears as expected when navigating to the Reader, Stats, Notifications, and the editor.
* For now, the banner's "open" link won't open the Jetpack app (see `deep links` section above for further details).  We should, however, confirm that the links to the WordPress app continue to function as before for users.

It'd also be useful to verify the banner works as expected with the following:   

* Different devices of varying sizes, including tablets.
* Different Calypso colour schemes.
* Different locales, including RTL. 

#### Screenshots

<details>
 <summary> LTR/RTL animation comparison ⤵️</summary>

| LTR  | RTL |
| ------------- | ------------- |
| <video src="https://user-images.githubusercontent.com/2998162/187277380-874c5103-deae-4c54-8f61-5bd1898a6973.mov" width="100%"> | <video src="https://user-images.githubusercontent.com/2998162/187278397-db082a5e-1bd4-4159-a6da-f39d0da842ea.mov" width="100%"> |

</details>

<details>
 <summary> Different sections ⤵️</summary>

| Reader  | Editor |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/187409623-21dcb9e4-07df-4121-8783-3927b914dc45.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/187408866-8e107c31-8b2c-428d-b8cb-15214e428f9e.png" width="100%"> |

| Stats  | Notes |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/187409734-2e097e6a-3eff-4e02-bf3f-3d6a593f988f.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/187413104-180b490b-7705-46f2-a6e2-5bb3f64b899c.png" width="100%"> |

</details>

<details>
 <summary> Different colour schemes ⤵️</summary>

| Midnight  | Sunrise |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/187413582-8bb97df0-ee6b-44f7-be2b-3a56495cb660.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/187413883-2b7811bd-96be-4243-9578-abd15c07d85d.png" width="100%"> |

</details>

<details>
 <summary> Mobile/tablet comparison  ⤵️</summary>

| iPad Air  | Landscape |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/187414032-93824625-d055-4285-bea7-44ee9f51c548.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/187414229-22b9a6fd-e7e3-418d-b41c-e00a4d1c1bfe.png" width="100%"> |

</details>

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

SiobhyB committed:
Create separate banners for the WP and JP apps

With this commit, two separate banners are created for both the WP and JP apps. The JP banner will be displayed when the 'jetpack/app-branding' feature flag is set to true (currently only for development builds and wpcalypso).

With this initial commit, both banners are the same as the current WP banner. The JP banner will be tweaked with future commits.

SiobhyB committed:
Clean up: readability spacing, delete stray return

SiobhyB committed:
Add lottie-web as a dependency, for animations

SiobhyB committed:
Update wording in banner's buttons for clarity

SiobhyB committed:
Remove circles in background from Jetpack banner

SiobhyB committed:
Add animated icon to Jetpack app banner

SiobhyB committed:
Replace 'react-lottie-player' for performance

By replacing the 'react-lottie-player' dependency with the 'official' 'lottie-web' dependency, it's hoped that there will be a positive impact on performance.

SiobhyB committed:
Use stripped back version of lottie dependency

SiobhyB committed:
Style Jerpack banner

The Jetpack-specific banner is styled via the use of a new '.jetpack' selector. This selector will eventually be removed when it's time for the Jetpack banner to be rolled out to users (along with the removal of any code related to the WordPress banner).

SiobhyB committed:
Rename component for clarity

Reposition component for overall readability

SiobhyB committed:
Update Android intents for Jetpack app

SiobhyB committed:
Remove additional, unwanted padding

SiobhyB commented:
👋 @griffbrad, non-urgent ping here. 

I'm currently working on an update to the app banner displayed on the mobile web. As per p9oQ9f-1jx-p2, I'm experimenting with using `lottie` to add an animation to the banner's icon. I'm using the most stripped back version of the library that's available, but still notice an increase of ~490% under [this comment's](https://github.com/Automattic/wp-calypso/pull/67013#issuecomment-1227698884) **Async-loaded Components** section.

I saw you mention at p9oQ9f-1jx-p2#comment-1985 that there shouldn't be as notable a difference if "you’re properly async loading the dependency", which leads me to believe I'm doing something wrong here. Would you perhaps be able to point me to any examples or tips for getting round this?  Forgive my ignorance, I've done some digging but it's not immediately clear to me what needs to be done.

Thanks in advance for any insight!

SiobhyB committed:
Add RTL versions of animated icons

griffbrad commented:
I should have been a bit more specific there: having that increase under “Async-loaded Components” is actually exactly what you’re aiming for! The concerning outcome would be the increase showing up under “App Entrypoints” or “Sections” because that would imply the code was being loaded for all users navigating to those sections rather than async when it was actually needed.

SiobhyB committed:
Merge branch 'trunk' into update/app-banner-branding

SiobhyB committed:
Styling tweaks to Jetpack app banner

SiobhyB committed:
Android deep links: Update intent URI for Jetpack

SiobhyB committed:
Merge branch 'trunk' into update/app-banner-branding

SiobhyB committed:
Fix broken unit tests by installing canvas

The unit test associated with the app banner was failing with the following error: "TypeError: Cannot set property 'fillStyle' of null"

As lottie-web uses canvas and jest doesn't handle canvas by default, it's necessary to install canvas as a dependency to prevent broken unit tests.

SiobhyB commented:
👋 @shaunandrews, would be grateful your review from a design perspective. All the screenshots are available under the PR's "Screenshots" section, let me know if anything extra is needed. I also wanted to note that the changes will be available on staging, for folks to test on actual devices, and I'll ask for a final approval in a separate PR before shipping to users at a later stage.

noahtallen commented on package.json:
```diff
@@ -219,6 +220,7 @@
 		"@wordpress/eslint-plugin": "^12.3.0",
 		"babel-loader": "^8.2.3",
 		"bunyan": "^1.8.15",
+		"canvas": "^2.9.3",
```
What do we use this for?

SiobhyB replied:
@noahtallen, I added this in https://github.com/Automattic/wp-calypso/pull/67013/commits/6c7a31f92090755d890f2e23e72834ac2e749a1f because the unit test associated with the banner was failing with the following: `TypeError: Cannot set property 'fillStyle' of null`

 I found that this is because `jest` isn't able to handle canvas by default (which is required by `lottie-web`), and requires a separate dependency for this purpose (similar issues found via Google [here](https://github.com/felippenardi/lottie-react-web/issues/21), [here](https://stackoverflow.com/questions/63219552/jest-import-statement-typeerror-cannot-set-property-fillstyle-of-null), and [here](https://robertmarshall.dev/blog/react-jest-tests-failing-using-lottie-js-solved/)). The [canvas dependency I added](https://www.npmjs.com/package/canvas) is an Automattic dependency.

Does this seem sound to you? Let me know if you have any further questions on this or if you have any concerns about this approach.

noahtallen replied:
Yep, that make sense to me! Thanks for clarifying :)

noahtallen commented on client/blocks/app-banner/utils.js:
```diff
@@ -13,36 +13,49 @@ export const ONE_MONTH_IN_MILLISECONDS = 2419200000; // 28 days
 				copy: translate(
 					'A streamlined editor with faster, simpler image uploading? Check and mate.'
 				),
+				icon: isRTL
+					? '/calypso/animations/app-promo/wp-to-jp-rtl.json'
+					: '/calypso/animations/app-promo/wp-to-jp.json',
```
```suggestion
				icon: `/calypso/animations/app-promo/wp-to-jp${ isRTL ? '-rtl' : '' }.json`
```

Just an idea to make this slightly less verbose :) (would also work below)

SiobhyB replied:
Ooh, nice! Thank you! I've gone ahead to implement your suggestion for all the lines in https://github.com/Automattic/wp-calypso/pull/67013/commits/0f6ba29bad9f7149d4476946f2557d584eb1aca7.

noahtallen commented on client/blocks/app-banner/index.jsx:
```diff
@@ -149,7 +157,47 @@ export class AppBanner extends Component {
 		const { title, copy } = getAppBannerData( translate, currentSection );
 
-		return (
+		const jetpackAppBanner = (
```
I might create two separate components (or functions) -- one for Jetpack, and one for WordPress. That way, we don't need to execute this code even when we aren't using it.

SiobhyB replied:
I like that idea, and have gone ahead to implement it in https://github.com/Automattic/wp-calypso/pull/67013/commits/e385f512e946afcf17c16034ebdcdfb4e690f0ac. It's worth noting that the code for the WordPress banner will eventually be deleted, so it may be worth revisiting this in the future when only the Jetpack banner exists.

noahtallen commented on client/blocks/app-banner/index.jsx:
```diff
@@ -190,9 +238,31 @@ export class AppBanner extends Component {
+export function BannerIcon( props ) {
+	const { translate, currentSection } = props;
+	const isRTL = useRtl();
+	const { icon } = getAppBannerData( translate, currentSection, isRTL );
```
Since we already have the data available in the parent component (line 158), what if we just passed it through props?

SiobhyB replied:
I've gone ahead to update this as per your suggestion, so that only the `icon` prop is being passed down:

https://github.com/Automattic/wp-calypso/blob/e385f512e946afcf17c16034ebdcdfb4e690f0ac/client/blocks/app-banner/index.jsx#L255-L267

Let me know if anything else is needed here!

noahtallen replied:
Nice! LGTM

noahtallen commented on client/blocks/app-banner/index.jsx:
```diff
@@ -1,9 +1,11 @@
-import { localize } from 'i18n-calypso';
+import { localize, useRtl } from 'i18n-calypso';
 import { get } from 'lodash';
+import lottie from 'lottie-web/build/player/lottie_light';
```
Do we use lottie for every single instance of `AppBanner`? If not, we might gain some performance by async loading it separately for just the instances which need it.

SiobhyB replied:
We will eventually (hopefully in a matter of weeks) be using the icon animation in all instances of the banner, so don't think there will be much gain from this, but let me know if you think otherwise.

noahtallen replied:
That sounds fair to me. :)

SiobhyB committed:
Make code more concise

SiobhyB committed:
Replace hook with HOC, for easier usage in class

To determine whether an RTL locale is in use, we're currently using the 'useRtl' hook. By replacing the hook with the 'withRtl' HOC, we're able to determine whether a locale is RTL directly within the class component, enabling 'isRTL' to be more cleanly passed down as a prop from the main banner component.

This commit removes the call to the 'useRtl' hook from BannerIcon (child) component. Its 'icon' prop will be defined in future commits.

SiobhyB committed:
Separate the Jetpack and WordPress banners

As it's not necessary to load the code for both the Jetpack and WordPress banners, this commit separates them out into separate functions. They'll now only be called when necessary.

noahtallen reviewed (approved):
The code changes LGTM -- I only tested one of the banners, but it worked well.

I think the main thing to consider moving forward is that this still adds almost 50KB just to the banners. Since we're on mobile, I think we can expect folks to have poor network connections sometimes, and that 50KB could make it more difficult for the banners to render expediently. That could theoretically cause a drop-off in the number of people clicking through the banner. 

Just wanted to throw the idea out there as you keep an eye on any stats related to this after deploying it! :)

I guess the bundle will immediately get bigger in production, even if the Jetpack changes aren't enabled yet. So you should be able to see if there's any impact pretty quickly :)

dcalhoun commented on client/blocks/app-banner/index.jsx:
```diff
@@ -190,9 +237,35 @@ export class AppBanner extends Component {
 
+function BannerIcon( { icon } ) {
+	useEffect( () => {
+		lottie.loadAnimation( {
+			container: document.querySelector( '.app-banner__icon' ),
+			renderer: 'svg',
+			loop: false,
+			autoplay: true,
+			path: icon,
+		} );
```
It seems unlikely a memory leak would occur, but it is probably best to destroy the animation in the Hook's clean-up function. 
```suggestion
		const animation = lottie.loadAnimation( {
			container: document.querySelector( '.app-banner__icon' ),
			renderer: 'svg',
			loop: false,
			autoplay: true,
			path: icon,
		} );
		return () => animation.destroy();
```

From looking at the detached nodes in the Chrome DevTools Memory tab, it did appear that destroying the animation caused the detached SVG elements to be released sooner. Again, I see no reason why the current code would cause a memory leak — the elements would likely still be released as no references to them are stored. However, cleaning up is likely still a safe practice worth implementing. 

| Before | After |
| - | - |
| <img width="569" alt="memory-before" src="https://user-images.githubusercontent.com/438664/187778438-3da385ca-62a6-47ca-9de2-42a1da7cb053.png"> | <img width="556" alt="memory-after" src="https://user-images.githubusercontent.com/438664/187778445-6d8face4-8611-480e-9120-f40a3ab6604f.png"> |

dcalhoun commented on client/blocks/app-banner/index.jsx:
```diff
@@ -190,9 +237,35 @@ export class AppBanner extends Component {
+			autoplay: true,
+			path: icon,
+		} );
+	}, [] );
```
To avoid unexpected outcomes and address linter warnings, we should probably list all the dependencies for this Hook. 
```suggestion
	}, [ icon ] );
```

dcalhoun commented on client/blocks/app-banner/index.jsx:
```diff
@@ -274,4 +347,4 @@ const mapDispatchToProps = {
 };
 
-export default connect( mapStateToProps, mapDispatchToProps )( localize( AppBanner ) );
+export default connect( mapStateToProps, mapDispatchToProps )( withRtl( localize( AppBanner ) ) );
```
Not a blocker by any means, but, if desired, we could leverage `connect` to avoid the rightward drift of the nested function transformations. The approach is found in other modules. 

https://github.com/Automattic/wp-calypso/blob/e385f512e946afcf17c16034ebdcdfb4e690f0ac/client/my-sites/activity/filterbar/date-range-selector.jsx#L206-L210

SiobhyB replied:
That looks so much cleaner! I've gone ahead to implement it in https://github.com/Automattic/wp-calypso/pull/67013/commits/4b4f68a8c39949db3c97196b96837e49eaf9bac3.

dcalhoun commented on client/blocks/app-banner/style.scss:
```diff
@@ -63,21 +69,46 @@ body.app-banner-is-visible {
+		/* stylelint-disable-next-line */
+		font-size: rem( 28px ); //typography-exception
+		font-weight: bold;
+		padding: 0 30px;
```
Rather than relying upon `padding`, I wonder if it might be more adaptable to set a `max-width` based upon the font size and use auto-margins. The parent element already has `padding`. Currently, the two containers combining padding means the text is quite narrow on small screens. This would apply to both the title and copy styles. WDYT? 

<details><summary>Small Screen Horizontal Spacing</summary>

| Before | After |
| - | - |
| <img width="321" alt="editor-before" src="https://user-images.githubusercontent.com/438664/187774582-1746a7de-40ae-416a-b310-f4d70e301da0.png"> | <img width="322" alt="editor-after" src="https://user-images.githubusercontent.com/438664/187774589-9248ac00-a018-4277-9669-5de4c3c463c6.png"> |
| <img width="324" alt="notifications-before" src="https://user-images.githubusercontent.com/438664/187774590-8862a7a9-0f6b-4128-83a8-ee5b9b074873.png"> | <img width="322" alt="notifications-after" src="https://user-images.githubusercontent.com/438664/187774595-9f1837f7-5e91-43a3-9d0a-f1b7f94ce0b9.png"> |

</details>

```css
.example-avoiding-long-lines {
  max-width: 18em;
  margin-left: auto;
  margin-right: auto;
} 
```

SiobhyB replied:
Thank you for the thorough testing here, David, you're right that that padding looks really off on smaller screens. I've gone ahead to implement your suggested update in https://github.com/Automattic/wp-calypso/pull/67013/commits/606c625eeedfaf674b3bb1be33d4a506783b462f and I think it works much better on smaller screens now. 🙇‍♀️

dcalhoun commented on package.json:
```diff
@@ -219,6 +220,7 @@
 		"@wordpress/eslint-plugin": "^12.3.0",
 		"babel-loader": "^8.2.3",
 		"bunyan": "^1.8.15",
+		"canvas": "^2.9.3",
```
> **If you have an ARM CPU**
>
> Prebuilds aren't available for ARM. Please see the [wiki](https://github.com/Automattic/node-canvas/wiki#installation-guides) for guides on building from source. See also https://github.com/node-gfx/node-canvas-prebuilt/issues/69.

This library does not appear to publish a prebuild for M1 Macs with ARM chips. This caused me issue when attempting to install this package. I ultimately installed additional dependencies so that building from source succeeded when running `yarn install`. Did you encounter this? 

<details><summary>Error Text</summary>

```
# This file contains the result of Yarn building a package (canvas@npm:2.9.3)
# Script name: install

node-pre-gyp info it worked if it ends with ok
node-pre-gyp info using node-pre-gyp@1.0.9
node-pre-gyp info using node@16.13.2 | darwin | arm64
node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.9.3/canvas-v2.9.3-node-v93-darwin-unknown-arm64.tar.gz
node-pre-gyp ERR! install response status 404 Not Found on https://github.com/Automattic/node-canvas/releases/download/v2.9.3/canvas-v2.9.3-node-v93-darwin-unknown-arm64.tar.gz
node-pre-gyp WARN Pre-built binaries not installable for canvas@2.9.3 and node@16.13.2 (node-v93 ABI, unknown) (falling back to source compile with node-gyp)
node-pre-gyp WARN Hit error response status 404 Not Found on https://github.com/Automattic/node-canvas/releases/download/v2.9.3/canvas-v2.9.3-node-v93-darwin-unknown-arm64.tar.gz
gyp info it worked if it ends with ok
gyp info using node-gyp@8.2.0
gyp info using node@16.13.2 | darwin | arm64
gyp info ok
gyp info it worked if it ends with ok
gyp info using node-gyp@8.2.0
gyp info using node@16.13.2 | darwin | arm64
gyp info find Python using Python version 3.10.6 found at "/opt/homebrew/opt/python@3.10/bin/python3.10"
gyp info spawn /opt/homebrew/opt/python@3.10/bin/python3.10
gyp info spawn args [
gyp info spawn args   '/Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/username/Library/Caches/node-gyp/16.13.2/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/username/Library/Caches/node-gyp/16.13.2',
gyp info spawn args   '-Dnode_gyp_dir=/Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/username/Library/Caches/node-gyp/16.13.2/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
Package cairo was not found in the pkg-config search path.
Perhaps you should add the directory containing `cairo.pc'
to the PKG_CONFIG_PATH environment variable
No package 'cairo' found
gyp: Call to 'pkg-config cairo --libs' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/lib/configure.js:353:16)
gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
gyp ERR! System Darwin 21.6.0
gyp ERR! command "/Users/username/.asdf/installs/nodejs/16.13.2/bin/node" "/Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--update-binary" "--module=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release/canvas.node" "--module_name=canvas" "--module_path=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release" "--napi_version=8" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v93"
gyp ERR! cwd /Users/username/Sites/a8c/wp-calypso/node_modules/canvas
gyp ERR! node -v v16.13.2
gyp ERR! node-gyp -v v8.2.0
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute '/Users/username/.asdf/installs/nodejs/16.13.2/bin/node /Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --update-binary --module=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release --napi_version=8 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v93' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/Users/username/Sites/a8c/wp-calypso/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1064:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
node-pre-gyp ERR! System Darwin 21.6.0
node-pre-gyp ERR! command "/Users/username/.asdf/installs/nodejs/16.13.2/bin/node" "/Users/username/Sites/a8c/wp-calypso/node_modules/@mapbox/node-pre-gyp/bin/node-pre-gyp" "install" "--fallback-to-build" "--update-binary"
node-pre-gyp ERR! cwd /Users/username/Sites/a8c/wp-calypso/node_modules/canvas
node-pre-gyp ERR! node -v v16.13.2
node-pre-gyp ERR! node-pre-gyp -v v1.0.9
node-pre-gyp ERR! not ok
Failed to execute '/Users/username/.asdf/installs/nodejs/16.13.2/bin/node /Users/username/Sites/a8c/wp-calypso/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --update-binary --module=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/Users/username/Sites/a8c/wp-calypso/node_modules/canvas/build/Release --napi_version=8 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v93' (1)
```

</details>

**Related Issues:**

* https://github.com/Automattic/node-canvas/issues/1511
* https://github.com/Automattic/node-canvas/issues/2078
* https://github.com/Automattic/node-canvas/issues/1662

Since we are only adding this for the Jest testing environment, I already wondered if [`jest-canvas-mock`](https://www.npmjs.com/package/jest-canvas-mock) might be a more appropriate package addition. However, these install errors may be another reason to switch. WDYT?

SiobhyB replied:
Oh, thanks for letting me know. 🤔 I didn't encounter the error you're seeing but the fact that there's so many mixed reports in those linked threads leads me to agree with replacing. You're right that `jest-canvas-mock` is likely more appropriate, too. I've gone ahead to replace the dependency in https://github.com/Automattic/wp-calypso/pull/67013/commits/d9ac75e4f9acdfe79d2cd75ce4d05ee7e68c6f1c.

dcalhoun reviewed (commented):
Great work! 🎉 I left a few comments to consider, but none are likely blockers. 

**Update:** I also meant to share I also tested all the banners, everything worked great. 👏🏻

SiobhyB committed:
List dependency for useEffect hook

To avoid any unexpected outcomes and address linter warnings, following feedback, this commit lists the dependency for the useEffect hook.

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Avoid possible memory leak by destroying animation

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Use 'compose' to group component wrappers

As the wrappers around the 'AppBanner' component were drifting far to the right and becoming hard to parse, this commit makes use of 'compose' to group all the wrappers together and improve readability.

SiobhyB committed:
Update text width/styling for smaller screens

The current padding in the banner's text elements causes short lines on smaller screens, which can looks off putting. This commit tweaks the CSS to remove the padding, instead adding a max-width to avoid longer lines (in addition to centring via margin: auto).

SiobhyB committed:
Replace 'canvas' with 'jest-canvas-mock'

As per this comment, it looks like there are some errors associated with the 'canvas' dependency: https://github.com/Automattic/wp-calypso/pull/67013#discussion_r959998759

We're also only adding this dependency for the Jest testing environment. As such, this commit replaces the use of 'canvas' with 'jest-canvas-mock'.

SiobhyB commented:
@noahtallen, @dcalhoun, thanks for your thorough reviews and guidance both. I believe I've addressed all your points, but let me know if anything more comes up!

> I think the main thing to consider moving forward is that this still adds almost 50KB just to the banners. Since we're on mobile, I think we can expect folks to have poor network connections sometimes, and that 50KB could make it more difficult for the banners to render expediently. That could theoretically cause a drop-off in the number of people clicking through the banner.

> I guess the bundle will immediately get bigger in production, even if the Jetpack changes aren't enabled yet. So you should be able to see if there's any impact pretty quickly :)

This is a really good point. If we do find a drop-off or problems in our own testing, it'd probably be worth reverting the animations and falling back to just a static icon.

As Shaun is AFK this week, perhaps the next step (pending any further review on the code) could be to merge this to staging. This would then make it easier to test the banners across different types of networks (as well as make it easier to test other elements, like deep linking). We could then make any updates that come from design feedback in a separate PR. Does that seem like a good route?

dcalhoun reviewed (approved):
🚀 Great work! 👏🏻 

One note: I was unable to get the banners to display on a tablet (emulated with Safari or Chrome developer tools). This was also the case for the production environment, so it appears unrelated to this PR. My guess is that we only display the banners for screens below a certain width. Did you experience this? 

> As Shaun is AFK this week, perhaps the next step (pending any further review on the code) could be to merge this so that the changes are available on staging. This would then make it easier to test the banners across different types of networks (as well as make it easier to test other elements, like deep linking). We could then keep an eye on any potentially negative impact and make any updates that come from design feedback in a separate PR. Does that seem like a good route?

This sounds like a good plan to me. 👍🏻

SiobhyB commented:
Thank you!

> One note: I was unable to get the banners to display on a tablet (emulated with Safari or Chrome developer tools). This was also the case for the production environment, so it appears unrelated to this PR. My guess is that we only display the banners for screens below a certain width. Did you experience this?

Ah, yes, I see you're right! While testing, I'd removed [the `shouldDisplayAppBanner` conditional](https://github.com/Automattic/wp-calypso/blob/d9ac75e4f9acdfe79d2cd75ce4d05ee7e68c6f1c/client/blocks/app-banner/index.jsx#L244-L246) locally so that the banner was always displayed to me, which is the only reason I was able to test with a tablet emulator. I see that, by default, the banner only displays on mobile devices and tablet testing isn't required. 👍 

I'll go ahead to merge now.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7507145

Thank you @SiobhyB for including a screenshot in the description! This is really helpful for our translators.

simison commented on client/blocks/app-banner/index.jsx:
```diff
@@ -190,9 +238,36 @@ export class AppBanner extends Component {
 
+function BannerIcon( { icon } ) {
+	useEffect( () => {
+		const animation = lottie.loadAnimation( {
```
Just learning about Lottie and I had a question: does the library take into account [`@media/prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) setting out of the box or would that need extra attention here?

dcalhoun replied:
My understanding is that, no, [Lottie does not take into account motion preferences](https://github.com/airbnb/lottie-web/issues/1986#issuecomment-586582516). 

You raise a good point. It's an opportunity to improve this new feature. I would imagine we should render the last frame of the animation when reduced motion is preferred.

SiobhyB replied:
Thanks for catching this! I'll work on a PR to take reduced animation preferences into account. 🙇‍♀️

SiobhyB replied:
To tie up all the loose ends, I've gone ahead to create that PR here: https://github.com/Automattic/wp-calypso/pull/67826

shaunandrews commented:
I tried a few times to get the banner to trigger, and was unsuccessful. However, since you did such an excellent job writing up this PR with a bunch of screenshots, I can still give some feedback!

The only note I have is about the font-weight. I see it's been set to `bold` for the headings, but it feels a little _too_ bold. Maybe we can go with `500` instead, and/or compare to other uses of Recoleta with bolding in our UI.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7507145).

SiobhyB commented:
@shaunandrews, thank you for the feedback! @dcalhoun also spotted some differences between the banner and the initial design you provided in a separate review, so I've gone ahead to create a separate (currently draft) PR with design updates at https://github.com/Automattic/wp-calypso/pull/67607. While that PR's in draft mode, I'll aim to keep the before/after screenshots up-to-date.

I've changed the title's `font-weight` from `bold` with `500` as part of that. I'll aim to set it as ready for review next week and will keep it up-to-date with any extra design feedback you may have. 🙇‍♀️
